### ISSUE TITLE
Add batch inference

### ISSUE THREAD

**Issue by Burhan-Q** [ultralytics/ultralytics#8721]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Related to PR #8058 

[Related comment](https://github.com/ultralytics/ultralytics/pull/8058?notification_referrer_id=NT_kwDOA7VQjLM5MzYyOTQ5MDQ3OjYyMjE0Mjg0#issuecomment-1980805191) from @glenn-jocher 

> ...yes some predictions sources will run in batches, but I think a main one that's missing is the glob or directory inference, though txt list of sources may also be missing. Yes please open an issue and tag us in it along with @adrianboguszewski from Intel. Thanks!

### Use case

> [!NOTE]  
> Currently batch inference is supported for certain types of input sources, this issue is to include the additional sources listed below

<details><summary>Example of existing batch inference support</summary>
<p>

```py
import cv2 as cv
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
im1 = "ultralytics/assets/bus.jpg"
im2 = "ultralytics/assets/zidane.jpg"

img1 = cv.imread(im1)
img2 = cv.imread(im2)

# Multi-image
results = model.predict(source=[img1, img2, img1, img1, img2, img1, img2, img2],)

>>> 0: 640x640 4 persons, 1 bus, 
1: 640x640 2 persons, 1 tie, 
2: 640x640 4 persons, 1 bus, 
3: 640x640 4 persons, 1 bus, 
4: 640x640 2 persons, 1 tie, 
5: 640x640 4 persons, 1 bus, 
6: 640x640 2 persons, 1 tie, 
7: 640x640 2 persons, 1 tie, 
155.0ms

[r.speed for r in results]

>>> [
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}, 
{'preprocess': 2.5001466274261475, 'inference': 19.374817609786987, 'postprocess': 10.336160659790039}
]

# Single image
results2 = model.predict(source=[img1,])

>>> 0: 640x480 4 persons, 1 bus, 1 stop sign, 192.1ms

results2[0].speed
>>> {'preprocess': 2.998828887939453, 'inference': 192.11935997009277, 'postprocess': 8.776426315307617}
```

</p>
</details> 

Add batch inference for (at least) the following sources:
- [ ] `glob` or directories
- [ ] list of text sources
- [ ] ...

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by tienhoang1994** [comment]
![image](https://github.com/ultralytics/ultralytics/assets/72494827/e56acbd9-de6d-4a7f-b0fc-2ad6c37744bd)
batch predict has no effect with my test like this (i test same video input so frame input already in same shape, but the document said that it auto set batchzise if i put list of image as input @glenn-jocher

**Comment by tienhoang1994** [comment]
i also tried with list.streams yolov8 provided, it also no improvement  in both speed and  hardware consuming

**Comment by Burhan-Q** [comment]
@tienhoang1994 you have 9 frames that complete processing in ~90 ms. Now you should test the time for inference on a single frame to compare like the example from the issue comment.

This issue was opened to track the work towards supporting additional batch inference sources. You can follow here to check progress but currently not all sources will support batch inference.

**Comment by glenn-jocher** [comment]
Thanks for the update, @tienhoang1994! 🚀 If you're seeing ~10ms for a single frame and ~90ms for 9 frames, it seems like the batch processing is indeed working as expected, offering a more efficient throughput compared to single-frame processing. The slight overhead might be due to initial setup or IO delays which can be amortized over larger batch sizes. If you have specific performance targets or further questions, feel free to share!

**Comment by Burhan-Q** [comment]
@tienhoang1994 the screenshot you shared shows inference for all 9 frames input as a batch. 
![image](https://github.com/ultralytics/ultralytics/assets/62214284/76d7dd7c-b5a5-4671-8a9e-4778ea91a49d)

The ~9-10 ms inference is for _each_ image in the batch. What I was saying and showed my initial comment is that the total batch inference time was `155.0ms` however when passing _only_ a single image, the inference time was `192.1ms`. If you see my [initial comment](https://github.com/ultralytics/ultralytics/issues/8721#issue-2171868759) and expand the section `► Example of existing batch inference support` to see the full details of the results I'm talking about

**Comment by tienhoang1994** [comment]
@Burhan-Q  ► Example of existing batch inference support  you mentioned about is exactly i expect. Total inference time batch of 8 imgs mostly same as 1 img(155 - 192ms).
![image](https://github.com/ultralytics/ultralytics/assets/72494827/973a98d7-8ec4-4c1a-9be7-ee6689a331cf)
i have tested again for you to see my result. it seem like batch not run all frame in parallel(~90ms for batch vs ~10ms for single). Thank you guys for your support, plz correct me if i have misunderstanding

**Comment by Burhan-Q** [comment]
@tienhoang1994 I follow your logic and results. I think maybe there is some subjectivity in the assessment of an appreciable difference. If I look at the initial batch results (your first comment) with a total of `~95ms` inference versus the looping result (you most recent comment) at `~115ms`, this is a difference of `20ms` and ~17% faster. It's not a lot, but to me, it's measurably faster.  

On my system, I think the difference is more noticeable, as inference on **one (1) image** was 192ms yet the 8-image batch inference was `155ms`. This processes 8x images in ~20% faster time than a single image. Rechecking, here are my results but using `timeit.repeat`:

## Code

```py
import timeit
import cv2 as cv
from ultralytics import YOLO
from functools import partial

model = YOLO("yolov8n.pt")

im1 = "ultralytics/assets/bus.jpg"
im2 = "ultralytics/assets/zidane.jpg"

img1 = cv.imread(im1)
img2 = cv.imread(im2)

p1 = partial(model.predict,(im1,)) # 1-image batch
p2 = partial(model.predict, [img1, img2, img1, img1, img2, img1, img2, img2]) # 8-image batch

timeit.repeat(p1, repeat=3, number=3)
timeit.repeat(p2, repeat=3, number=3)
```

## 1 image batch inference
<details><summary>screenshot 1 image batch</summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/5b6153dd-b891-41d1-b0a5-49642d91089b)

</p>
</details> 

- first iteration inference time is `141ms`
- average of remaining 7 repeats is `~8ms`

## 8 image batch inference

<details><summary>screenshot 8 image batch</summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/d91e6add-b6cb-4b1a-8627-ce518c3e0db7)

</p>
</details> 


- first iteration inference time is `152ms`
- average of remaining 7 repeats is `~15ms`

Ignoring the initial "slow" (warmup) result, the 8-image batch takes just under 2x longer than the 1-image batch in my recent test, but it is also processing 8x input data. Calculating the inference speed of _individual_ images in the 8-image batch, I'll use the final 8-image batch inference time, I'll divide the total inference time (14.6ms) by the number of images (8) which is (1.8ms) per image for batch inference. Looking at the (last entry) 1-image batch inference time (8.6ms), you can see that the per-image inference time for the batch of 8 images is _much faster_.

**Comment by tienhoang1994** [comment]
sorry for wasting your time. maybe i found the problem is about GPU.
![image](https://github.com/ultralytics/ultralytics/assets/72494827/905e3696-79ff-4b95-9962-f56ffc05eabf)
on the left - weak GPU 1060 i am using on this Post has 7ms per img vs 5.6ms per img in batch of 8
on the right - strong GPU 3090 has 5.1ms per img vs 1.1ms per img in batch of 8(much better)

**Comment by glenn-jocher** [comment]
@tienhoang1994 hey there! No worries at all, you're not wasting our time. We're here to help! 😊 It looks like you've made an interesting observation regarding the impact of GPU capabilities on batch inference performance. Indeed, the difference in processing power between a GTX 1060 and an RTX 3090 can significantly affect the efficiency gains from batch processing. 

The RTX 3090, with its higher compute capability and memory bandwidth, can better leverage parallel processing, making the advantages of batch inference more pronounced. On the other hand, the GTX 1060, while still a capable GPU, might not exhibit as dramatic improvements due to its hardware limitations.

Here's a quick example to illustrate how you might adjust batch sizes based on your GPU's capabilities:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.pt')

# Define your source
source = ['path/to/image1.jpg', 'path/to/image2.jpg']  # and so on...

# Predict with batch inference
results = model.predict(source=source)
```

Remember, finding the optimal batch size for your specific hardware setup can maximize your inference efficiency. Keep experimenting, and thanks for sharing your findings! 👍

**Comment by Burhan-Q** [comment]
Additional batch inferencing included with #8817
